src/libsyntax/parse/parser.rs

```diff
@@ -3093,8 +3093,7 @@ impl<'a> Parser<'a> {
         self.expect(&token::FatArrow)?;
         let expr = self.parse_expr_res(Restrictions::RESTRICTION_STMT_EXPR, None)?;
 
-        let require_comma =
-            !classify::expr_is_simple_block(&expr)
+        let require_comma = classify::expr_requires_semi_to_be_stmt(&expr)
```

Comment:
>A copy of fn expr_requires_semi_to_be_stmt in librustc/hir/print.rs is also badly out-of-date.

Is it? It seems legit, and even already handles catch (catch is just a block in hit, isn't it?)